### ISSUE TITLE
Cannot read memory asynchronously, between processes

### ISSUE THREAD

**Issue by funatsufumiya** [kyr0/libsharedmemory#3]
I tested with different processes, but read data won't change.  Maybe while other process is writing something. I think some kind of mutex or double buffer is required here.

**Comment by kyr0** [comment]
Hmm, yeah, there are two cases here:

a) the change flag is changed by two writer processes. Because this acts as a toggle, it can create a situation where reader processes don't recognize changes. The data might change but the code doesn't "identify" the change. This could be solved by replacing the bit with a CRC checksum and checking the checksum for change.

b) two writer processes are writing and overwriting themselves. If the data is "the same", and readers "lack behind" in timing, they wouldn't see the change that happened "in between". Adding a revision byte that counts up would help here to identify "missed changes". If the counter increased > +1 between the last read and the current read, the case is clear. However this wouldn't solve the "What's the data that I've missed?" Question. To implement this, the library could allocate `n` more shared memory buffers (configurable); these would start with the index of the counter number (index), followed by the data type (see readme) and then the data. If the library detects that the counter has "jumped" more than +1 change, it could find the right data in those buffers, by checking the buffers, looking for the n+1 index.

I don't have much time these days, but I'd welcome PR's if you like to play with this maybe?

**Comment by kyr0** [comment]
@funatsufumiya Had a few more thoughts... RACE was still possible with the previous idea; I edited my comment :)

**Comment by funatsufumiya** [comment]
@kyr0 
Thank you for your prompt reply!

I was trying to use it like the so-called PUSH/PULL or PUB/SUB, but I was struggling because the data was only loaded correctly the first time. 

The first thing I thought of was to create a couple of shared memory locations with similar names and swap them with a separate area that points to which one is being used, but I will try the method you adviced me :)

**Comment by funatsufumiya** [comment]
I also had an idea to create an area that corresponds to a mutex lock, and both the reader and writer use lock there, but it seemed to cause problems such as the locks not being removed permanently, so I would like to try various methods.

**Comment by funatsufumiya** [comment]
I just came up with the idea for PUSH/PULL style, that the shared memory should have a different (incremental) name each time and the reader should delete the memory once it has been read. That way, the write side would always write to a new area and not have to know the behavior of the read side, which may be the best way for my current needs.

**Comment by kyr0** [comment]
Yep, that would resemble parts of the idea of process_ids basically; you can also simply create a temporary file where the process ids of writers are named in. If the name of the shared memory is prefixed with the process id, the reader only need to watch those shared memory locations. Now to identify what's the most recent data, you can simply use a few bytes *before* the data (prefix) encoding a timestamp. The reader then uses the data with the most recent timestamp :)

**Comment by funatsufumiya** [comment]
@kyr0 Thank you! I'll try them!

**Comment by funatsufumiya** [comment]
@kyr0
Upon further investigation, I found that there are already several high-level IPC Messaging implementations like the one I am looking for, and here are links to representative ones.

- [shadesmar](https://github.com/Squadrick/shadesmar)
- [alephzero](https://github.com/alephzero/alephzero)
- [iceoryx](https://github.com/eclipse-iceoryx/iceoryx)

(And also, but much higher level, may you already know)

- [nanomsg](https://github.com/nanomsg/nanomsg)
- [NNG](https://github.com/nanomsg/nng)
- [ZeroMQ](https://github.com/zeromq)

Referenced Issue:
https://github.com/Squadrick/shadesmar/issues/56

**Comment by funatsufumiya** [comment]
By the way, if we look at it not as a messaging protocol but in the direction of Ephemeral DB, I think [SimDB](https://github.com/LiveAsynchronousVisualizedArchitecture/simdb/) (and many more) would be the category of high-level implementations.

**Comment by funatsufumiya** [comment]
I couldn't find this before, but [cpp-ipc](https://github.com/mutouyun/cpp-ipc) seems to be an IPC implementation with a simple mechanism like PUBSUB. (The number of receivers seems to be limited to 32.)

https://github.com/mutouyun/cpp-ipc

**Comment by funatsufumiya** [comment]
Looking at `cpp-ipc`, I'm starting to wonder if `libsharedmemory` itself could have a simple channel mechanism (otherwise anyone would be reinventing the wheel as a result), so I'll re-open the Issue itself in the meantime and see if a PR can be created.

**Comment by kyr0** [comment]
Oh thank you for all the research you've done. I think we could add those findings to the README.md as well. A section line "Alternatives" could be helpful for others I guess? 

And yeah, there's a lot of alternatives out there already -- I personally think that simplicity is key for good software quality. The most simple solution that does the job is often the best IMHO.. it allows code to be understood by anyone, adapted, maintained, fixed etc. pp. That's why I sometimes reinvent the wheel on purpose.. just to get to a solution that is waaay simpler than the madness of overcomplexity that we can often find out there :)

**Comment by funatsufumiya** [comment]
I generally agree, but in this case, many cases need asynchronous use from multiple applications, so I wonder that some simple asynchronous support would be good.

**Comment by kyr0** [comment]
@funatsufumiya  Fixed with v2.0.0 :)

**Comment by funatsufumiya** [comment]
Thank you. I confirmed inter-process problem was fixed in v2.0.0! Really awesome job.

My openFrameworks v0.12.0 (oF) used C++17 and `using namespace std` was default used on oF, this makes problems (ambiguous `byte`, `std::span` not defined, C++20 functions cannot be used etc) on Windows VS2022, so I little tweaked on my oF addon [src](https://github.com/funatsufumiya/ofxSharedMemory/tree/main/src) for C++11 ~ C++17. I checked this is also working on UNIX (mac) C++17 environment.
